### ISSUE TITLE
Issue with mpgen and RUNPATH

### ISSUE THREAD

**Issue by fanquake** [bitcoin/bitcoin#19981]
Pointed out by @ryanofsky in [#19685](https://github.com/bitcoin/bitcoin/pull/19685#issuecomment-692861142):

> Still confirming but it seems like this PR might be causing an error on travis: "/home/travis/build/bitcoin/bitcoin/depends/x86_64-pc-linux-gnu/share/../native/bin/mpgen: error while loading shared libraries: libcapnpc-0.7.0.so: cannot open shared object file: No such file or directory" error in https://travis-ci.org/github/bitcoin/bitcoin/jobs/726983595#L2545

> Running CI scripts locally, I'm able to reproduce the error. But if I revert the changes in this PR the error goes away. The difference seems to be caused by cmake no longer setting RUNPATH after this PR. With this PR, RUNPATH is empty, but if this PR is reverted, readelf shows:

```bash
$ readelf -d depends/x86_64-pc-linux-gnu/native/bin/mpgen | grep RUNPATH
 0x000000000000001d (RUNPATH)            Library runpath: [/home/russ/src/bitcoin/depends/x86_64-pc-linux-gnu/native/lib]
```

> I need to dig in more but two questions I have are:

> 1. Why did travis not catch this problem when this PR was being worked on or after this PR was merged. Is there a cache that we failed to clear? Is there a cache we should be clearing when we update depends scripts in the future? (The travis error above started happening after I pushed a PR update that bumped the `depends/packages/native_libmultiprocess.mk` package, so I think that accounts for why a cache might have been avoided in the failing run above).

> 2. What change in this PR might be clobbering the cmake RUNPATH? If I had to guess it might be happening because the LDFLAGS environment variable was unset previously, but now it is set

**Comment by ryanofsky** [comment]
Here is what I've been able to figure out with this issue so far.

1. I don't think there's a travis caching problem. Travis passing on master but failing in the 726983595 build is just a result of travis job 726983595 above executing `mpgen` in the bitcoin build and current master not doing that. Any changes to `funcs.mk` changes are also reflected in build ids so if build ids are used correctly, there probably shouldn't be a caching issue.

2. Confirmed setting LDFLAGS variable is the thing that causes RUNPATH not to be set. LDFLAGS environment value gets copied into CMAKE_EXE_LINKER_FLAGS cached global value, which otherwise would be empty. So prior to #19685, LDFLAGS/CMAKE_EXE_LINKER_FLAGS value was empty. After #19685 it is "-L/home/russ/src/bitcoin/depends/x86_64-pc-linux-gnu/native/lib"

3. So far I don't see a clean fix for this. Partially reverting #19685 and not passing LDFLAGS for native packages is probably not ideal. Overriding LDFLAGS for native_libmultiprocess in particular is hacky. I haven't figured out exactly how setting LDFLAGS does prevent mpgen RUNPATH from being set. MPGEN target has [`INSTALL_RPATH_USE_LINK_PATH`](https://github.com/chaincodelabs/libmultiprocess/blob/4c599773923afe6829139ce682d723d769e335dc/CMakeLists.txt#L82) set to true. If LDFLAGS is empty or set to harmless value like `-lm` with no `-L` link path, RUNPATH is set. But as soon as `-L` is added, RUNPATH disappears. I think this might have to do with logic https://gitlab.kitware.com/cmake/cmake/-/blob/11425041f04fd0945480b8f9e9933d1549b93981/Source/cmComputeLinkInformation.cxx#L1720, but so far I don't understand it. I might experiment with LIBRARY_PATH environment variable, but even if this works it would need to be set for the native packages, unset for native packages, so would add unappealing complication. 

4. Fastest way to test this is to hardcode capnp package build id so it doesn't get rebuilt with every change to `funcs.mk`

```diff
--- a/depends/funcs.mk
+++ b/depends/funcs.mk
@@ -45,6 +45,7 @@ $(eval $(1)_all_dependencies:=$(call int_get_all_dependencies,$(1),$($($(1)_type
 $(foreach dep,$($(1)_all_dependencies),$(eval $(1)_build_id_deps+=$(dep)-$($(dep)_version)-$($(dep)_recipe_hash)))
 $(eval $(1)_build_id_long:=$(1)-$($(1)_version)-$($(1)_recipe_hash)-$(release_type) $($(1)_build_id_deps) $($($(1)_type)_id_string))
 $(eval $(1)_build_id:=$(shell echo -n "$($(1)_build_id_long)" | $(build_SHA256SUM) | cut -c-$(HASH_LENGTH)))
+$(eval native_capnp_build_id:=f5a3acc1fc3)
 final_build_id_long+=$($(package)_build_id_long)
 
 #compute package-specific paths
```

After that resulting RUNPATH can be checked with:

```sh
rm -rvf depends/x86_64-pc-linux-gnu/native depends/work/staging depends/work/build
make -C depends MULTIPROCESS=1 native_libmultiprocess_staged
for f in `find -name mpgen`; do echo == $f ==; readelf -d $f | grep -i path; done
```

CMakeCache can be examined with:

```sh
make -C depends MULTIPROCESS=1 native_libmultiprocess_built
find -name CMakeCache.txt
```

**Comment by ryanofsky** [comment]
I think I found a decent workaround setting CMAKE_INSTALL_RPATH. If it works I will incorporate it in #19160 and close this issue. Setting LIBRARY_PATH environment variable didn't have any effect (was just ignored).

Also posted a question about this upstream https://discourse.cmake.org/t/install-rpath-use-link-path-not-working-when-cmake-exe-linker-flags-ldflags-is-set/1892

**Comment by ryanofsky** [comment]
Will close this issue. I confirmed 01a7b773606c7304e5ca42f6ff121d818c972361 from #19160 fixes the problem, or at least works around it: https://travis-ci.org/github/bitcoin/bitcoin/jobs/730006938.

I could make a separate PR for the fix since it's a build change pretty different than the other changes in #19160. On the other hand, a reason for keeping it in the same PR is that other changes in the PR are needed to verify the fix works on travis. Maybe it would be a good idea in the future for the depends build to have a test target (particularly for native packages, but checks could be done on cross-compiled packages too), so build fixes like 01a7b773606c7304e5ca42f6ff121d818c972361 could be more separable and testable.
